File: system-design/systems/01-rate-limiter/java/SlidingWindowCounter.java
import java.time.Duration;
import java.util.function.LongSupplier;

/** 用前一固定窗口的线性权重，近似统计最近一个窗口内的请求量。 */
public class SlidingWindowCounter {
    private final double limit;
    private final long windowNanos;
    private final LongSupplier nowNanos;
    private double previousCount;
    private double currentCount;
    private long currentWindowStart;

    public SlidingWindowCounter(double limit, Duration window) {
        this(limit, window, System::nanoTime);
    }

    public SlidingWindowCounter(double limit, Duration window, LongSupplier nowNanos) {
        if (!Double.isFinite(limit) || limit <= 0) {
            throw new IllegalArgumentException("limit must be finite and greater than zero");
        }
        if (window == null || window.isZero() || window.isNegative()) {
            throw new IllegalArgumentException("window must be greater than zero");
        }
        if (nowNanos == null) {
            throw new IllegalArgumentException("nowNanos must not be null");
        }

        long nanos;
        try {
            nanos = window.toNanos();
        } catch (ArithmeticException exc) {
            throw new IllegalArgumentException("window is too large", exc);
        }
        if (nanos <= 0) {
            throw new IllegalArgumentException("window must be at least one nanosecond");
        }

        this.limit = limit;
        this.windowNanos = nanos;
        this.nowNanos = nowNanos;
        this.currentWindowStart = nowNanos.getAsLong();
    }

    public synchronized boolean allow() {
        return allow(1);
    }

    public synchronized boolean allow(double n) {
        if (!Double.isFinite(n) || n <= 0) {
            return false;
        }

        long t = nowNanos.getAsLong();
        long elapsed = t > currentWindowStart ? t - currentWindowStart : 0;
        long windowsElapsed = elapsed / windowNanos;
        if (windowsElapsed == 1) {
            previousCount = currentCount;
            currentCount = 0;
        } else if (windowsElapsed >= 2) {
            previousCount = 0;
            currentCount = 0;
        }

        if (windowsElapsed >= 1) {
            currentWindowStart += windowsElapsed * windowNanos;
            elapsed = t - currentWindowStart;
        }

        double previousWeight = 1.0 - (double) elapsed / windowNanos;
        double estimate = currentCount + previousCount * previousWeight;
        if (estimate + n > limit) {
            return false;
        }
        currentCount += n;
        return true;
    }
}

File: system-design/systems/01-rate-limiter/python/sliding_window_counter.py
import math
import threading
import time
from typing import Callable


class SlidingWindowCounter:
    """用前一固定窗口的线性权重，近似统计最近一个窗口内的请求量。"""

    def __init__(
        self,
        limit: float,
        window_seconds: float,
        now: Callable[[], float] = time.monotonic,
    ) -> None:
        try:
            limit = float(limit)
            window_seconds = float(window_seconds)
        except (TypeError, ValueError) as exc:
            raise ValueError("limit and window_seconds must be numbers") from exc
        if not math.isfinite(limit) or limit <= 0:
            raise ValueError("limit must be finite and greater than zero")
        if not math.isfinite(window_seconds) or window_seconds <= 0:
            raise ValueError("window_seconds must be finite and greater than zero")
        if not callable(now):
            raise ValueError("now must be callable")

        self.limit = limit
        self.window_seconds = window_seconds
        self._now = now
        self._previous_count = 0.0
        self._current_count = 0.0
        self._current_window_start = now()
        self._lock = threading.Lock()

    def allow(self, n: float = 1.0) -> bool:
        try:
            n = float(n)
        except (TypeError, ValueError):
            return False
        if not math.isfinite(n) or n <= 0:
            return False

        with self._lock:
            t = self._now()
            elapsed = max(0.0, t - self._current_window_start)
            windows_elapsed = int(elapsed // self.window_seconds)
            if windows_elapsed == 1:
                self._previous_count = self._current_count
                self._current_count = 0.0
            elif windows_elapsed >= 2:
                self._previous_count = 0.0
                self._current_count = 0.0

            if windows_elapsed >= 1:
                self._current_window_start += windows_elapsed * self.window_seconds
                elapsed = max(0.0, t - self._current_window_start)

            previous_weight = 1.0 - elapsed / self.window_seconds
            estimate = self._current_count + self._previous_count * previous_weight
            if estimate + n > self.limit:
                return False
            self._current_count += n
            return True
